api/v1beta1/openstackmachine_types.go

```diff
@@ -33,6 +33,60 @@ const (
 	IPClaimMachineFinalizer = "openstackmachine.infrastructure.cluster.x-k8s.io/ip-claim"
 )
 
+// SchedulerHintValueType is the type that represents allowed values for the Type field.
+// +kubebuilder:validation:Enum=Bool;String;Number
+type SchedulerHintValueType string
+
+// Constants representing the allowed types for SchedulerHintAdditionalValue.
+const (
+	SchedulerHintTypeBool   SchedulerHintValueType = "Bool"
+	SchedulerHintTypeString SchedulerHintValueType = "String"
+	SchedulerHintTypeNumber SchedulerHintValueType = "Number"
+)
+
+// SchedulerHintAdditionalValue represents the value of a scheduler hint property.
+// The value can be of various types: Bool, String, or Number.
+// The Type field indicates the type of the value being used.
+// +kubebuilder:validation:XValidation:rule="has(self.type) && self.type == 'Bool' ? has(self.bool) : !has(self.bool)",message="bool is required when type is Bool, and forbidden otherwise"
+// +kubebuilder:validation:XValidation:rule="has(self.type) && self.type == 'Number' ? has(self.number) : !has(self.number)",message="number is required when type is Number, and forbidden otherwise"
+// +kubebuilder:validation:XValidation:rule="has(self.type) && self.type == 'String' ? has(self.string) : !has(self.string)",message="string is required when type is String, and forbidden otherwise"
+// +union.
+type SchedulerHintAdditionalValue struct {
+	// Type represents the type of the value (Bool, String, or Number).
```

Comment:
Should explain ideally that the valid values are `Bool`, `String` and `Number`. Your statement includes them but it would be better to be direct and explicit
```suggestion
	// Type represents the type of the value.
	// Valid values are Bool, String, and Number.
```